### PR TITLE
#185448076 : Governance rules support

### PR DESCRIPTION
**Pull request by xinghengwang**
- Before the styling isn't consistent some files are 4 space indent and some are 2 spaces indent. So I added rubocop style guidelines, and it auto formatted a lot of files. 
- I also wrote an automated test framework to help testing. SDK constributors please chat with me regarding how to use it. 
  https://github.com/Moesif/governance-rules-tests

But please review before merging. lets squash and merge.

### CODE REVIEW

**Comment by praves77** [comment]
no worries, people can ignore whitespaces while looking into diff for review..

File: .rubocop.yml
Comment (praves77) [inline .rubocop.yml:1]: Nice

File: Gemfile.lock [hunk 1]
```diff
@@ -1,51 +1,49 @@
...
       power_assert
     unf (0.1.4)
       unf_ext
-    unf_ext (0.0.8)
+    unf_ext (0.0.8.2)
 
 PLATFORMS
   arm64-darwin-21
-  ruby
-  x86_64-linux
```
Comment (praves77) [inline Gemfile.lock]: @xinghengwang any reason for removing these 2 platforms?

File: lib/moesif_rack/app_config.rb [hunk 2]
```diff
@@ -3,134 +3,101 @@
...
-                return 100
-            end
-        rescue => exception
-            @moesif_helpers.log_debug 'Error while geting sampling percentage, assuming default behavior'
-            @moesif_helpers.log_debug exception.to_s
-            return 100
-        end
-    end
+    # Check if response body is not nil
+    return response_body, config_api_response.headers[:x_moesif_config_etag], Time.now.utc unless response_body.nil?
```
Comment (praves77) [inline lib/moesif_rack/app_config.rb:41]: ah.. it's syntax matches with coffeescript

File: lib/moesif_rack/client_ip.rb
Comment (praves77) [inline lib/moesif_rack/client_ip.rb:1]: did it run lint on lib too? or it's different version lib?

File: Gemfile.lock (hunk 1 above)
Comment (xinghengwang) [reply Gemfile.lock]: not sure. seems auto generated. let me double check,

File: lib/moesif_rack/client_ip.rb
Comment (xinghengwang) [reply lib/moesif_rack/client_ip.rb:1]: hm.. I just formatted this file. this file is not from another lib. this is part of this repo.

File: lib/moesif_rack/moesif_middleware.rb [hunk 3]
```diff
@@ -32,38 +31,40 @@ def initialize app, options = {}
       @debug = options['debug']
       @app_config = AppConfig.new(@debug)
       @moesif_helpers = MoesifHelpers.new(@debug)
-      @config = @app_config.get_config(@api_controller)
+      # @config = @app_config.get_config(@api_controller)
       @config_etag = nil
```
Comment (xinghengwang) [inline lib/moesif_rack/moesif_middleware.rb:35]: This load is uneeded because immediately on line 52 it load again.

**Comment by praves77** [comment]
It looks fine to me. I will review again later tonight or tomorrow morning then we can publish.

@XinyuLuan can you please review the test cases and apply rule? Thanks.

File: lib/moesif_rack/governance_rules.rb [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+    unless config_user_rules_values.nil?
+      config_user_rules_values.each do |entry|
+        rule_id = entry['rules']
+        # this is user_id matched cohort set in the rule.
+        rule_ids_hash_that_is_in_cohort[rule_id] = true unless rule_id.nil?
+        # rule_ids_hash_that_I_am_in_cohot{629847be77e75b13635aa868: true}
+
+        found_rule = @user_rules[rule_id]
+        if found_rule.nil?
+          @moesif_helpers.log_debug('rule for not foun for ' + rule_id.to_s)
```
Comment (HzANut) [inline lib/moesif_rack/governance_rules.rb:259]: @xinghengwang Just a nitpic, here is a typo for "found"

**Review by HzANut — COMMENTED** [review]
lgtm

**Pull request merged**